### DISCUSSION TITLE
Add queryWithBinds to Selector layer

### DISCUSSION THREAD

**Discussion by MayTheSForceBeWithYou** [apex-enterprise-patterns/fflib-apex-common#481] [General]
In Winter '24 release, Salesforce introduced the ability to make SOQL queries with bound parameters:
[Dynamically Pass Bind Variables to a SOQL Query](https://help.salesforce.com/s/articleView?id=release-notes.rn_apex_bind_var_soql.htm&release=242&type=5). Add this functionality to the fflib Selector layer.

**Describe the solution you'd like**
// TODO: A clear and concise description of what you want to happen.

**Describe alternatives you've considered**
// TODO: A clear and concise description of any alternative solutions or features you've considered.

**Additional context**
// TODO: Add any other context or screenshots about the feature request here.

**Comment by wimvelzeboer** [comment]
It might be a way to make real unit-test (with mocking) for selector classes. Now we can't really make those tests without going to the database.

**Comment by daveespo** [comment]
@MayTheSForceBeWithYou  -- Hi Nathan -- hope all is well 

Can you elaborate a bit more with a couple of example use cases? i.e. pseudocode for how you'd like a feature like this to work?

`fflib_SObjectSelector` does not actually "run the query" (disregarding the two special case `selectSObjectsById` and `queryLocatorById` default 'finder' methods) -- it's up to the author of the Selector class to do `Database.query` .. and that same author could use `Database.queryWithBinds` today without any further enhancements to `fflib_SObjectSelector`

**Comment by pgajek2** [comment]
The [SOQL Lib](https://soql.beyondthecloud.dev/) has `Database.queryWithBinds`, so you can bind variables even in place of usage. 
There is no need to `Database.query` in the selector class. 

```java
@AuraEnabled
public static List<Account> getPartnerAccounts(String accountName) {
   return AccountsSelector.query()
      .byRecordType('Partner')
      .whereAre(SOQL.Filter.name().contains(accountName)) // < == variable automatically binded
      .with(Account.BillingCity, Account.BillingCountry)
      .toList();
 }
```

```java
public inherited sharing class AccountsSelector extends SOQL implements SOQL.Selector {
    public static final String MOCK_ID = 'SOQL_Account';

    public static SOQL_Account query() {
        return new SOQL_Account();
    }

    private SOQL_Account() {
        super(Account.SObjectType);
        // default settings
        with(Account.Id, Account.Name);
        systemMode();
        withoutSharing();
        mockId(MOCK_ID);
    }

    public SOQL_Account byType(String type) {
        whereAre(Filter.with(Account.Type).equal(type));
        return this;
    }

    public SOQL_Account byIndustry(String industry) {
        whereAre(Filter.with(Account.Industry).equal(industry));
        return this;
    }

    public SOQL_Account byParentId(Id parentId) {
        whereAre(Filter.with(Account.ParentId).equal(parentId));
        return this;
    }

    public SOQL_Account byOwnerId(Id ownerId) {
        whereAre(Filter.with(Account.OwnerId).equal(ownerId));
        return this;
    }
}
```
